common/fixation_entry.go

```diff
@@ -61,48 +71,106 @@ type FixationStore struct {
+		ASCII_DEL = 127 // ascii for DEL
+	)
+
+	for i := 0; i < len(index); i++ {
+		if index[i] < ASCII_MIN || index[i] > ASCII_MAX {
+			return index, types.ErrInvalidIndex
+		}
+	}
+	return index + string([]byte{ASCII_DEL}), nil
+}
+
+// desantizeIndex reverts the effect of sanitizeIndex - removes the trailing
+// (ascii) DEL terminator.
+func desanitizeIndex(safeIndex string) string {
+	return safeIndex[0 : len(safeIndex)-1]
+}
+
+func (fs *FixationStore) getStore(ctx sdk.Context, index string) *prefix.Store {
+	store := prefix.NewStore(
+		ctx.KVStore(fs.storeKey),
+		types.KeyPrefix(fs.createStoreKey(index)))
+	return &store
+}
+
+// setEntry modifies an existing entry in the store
+func (fs *FixationStore) setEntry(ctx sdk.Context, entry types.Entry) {
+	store := fs.getStore(ctx, entry.Index)
+	byteKey := types.EncodeKey(entry.Block)
+	marshaledEntry := fs.cdc.MustMarshal(&entry)
+	store.Set(byteKey, marshaledEntry)
+}
+
 // AppendEntry adds a new entry to the store
 func (fs *FixationStore) AppendEntry(ctx sdk.Context, index string, block uint64, entryData codec.ProtoMarshaler) error {
-	// get the latest entry for this index
-	latestEntry, found := fs.getUnmarshaledEntryForBlock(ctx, index, block)
+	safeIndex, err := sanitizeIndex(index)
+	if err != nil {
+		details := map[string]string{"index": index}
+		return utils.LavaError(ctx, ctx.Logger(), "AppendEntry_invalid_index", details, "invalid non-ascii entry")
+	}
+
+	latestEntry, found := fs.getUnmarshaledEntryForBlock(ctx, safeIndex, block)
 
 	// if latest entry is not found, this is a first version entry
 	if !found {
-		fs.SetEntryIndex(ctx, index)
+		fs.setEntryIndex(ctx, safeIndex)
 	} else {
 		// make sure the new entry's block is not smaller than the latest entry's block
 		if block < latestEntry.GetBlock() {
-			return utils.LavaError(ctx, ctx.Logger(), "AppendEntry_block_too_early", map[string]string{"latestEntryBlock": strconv.FormatUint(latestEntry.GetBlock(), 10), "block": strconv.FormatUint(block, 10), "index": index, "fs.prefix": fs.prefix}, "can't append entry, earlier than the latest entry")
+			details := map[string]string{
+				"latestBlock": strconv.FormatUint(latestEntry.GetBlock(), 10),
+				"block":       strconv.FormatUint(block, 10),
+				"index":       index,
+				"fs.prefix":   fs.prefix,
+			}
+			return utils.LavaError(ctx, ctx.Logger(), "AppendEntry_block_too_early", details, "entry block earlier than latest entry")
 		}
 
 		// if the new entry's block is equal to the latest entry, overwrite the latest entry
 		if block == latestEntry.GetBlock() {
 			return fs.ModifyEntry(ctx, index, block, entryData)
 		}
+
+		// if the old latest entry has refcount of 0, then update its "stale_at" time
+		// TODO: remove this when the latest entry gets its own refcount.
+		if latestEntry.Refcount == 0 {
+			// never overflows because because ctx.BlockHeight is int64
+			latestEntry.StaleAt = uint64(ctx.BlockHeight()) + uint64(types.STALE_ENTRY_TIME)
```

Comment:
This check is for the entry that was "latest entry" just prior to appending the current one. So StaleAt could not have been set for this entry - it was in use by definition, due to being the latest.
If nobody else took reference to it (e.g. either no-one did GetPlan() to got the latest plan, or someone did but later also called PutPlan() on it) - then it is unneeded from exactly now, so we set the StaleAt.
If there is a reference to it, because it is in use by someone, then the ref-count won't be zero and we won't set StaleAt at all.